Unet3d model bring up start: working ttnn/reference impl and passing …

mert-kurttutan opened the pull request:
### Ticket
Linked issue #30858


### Problem description
Model bring up for 3D DenseUNet

### What's changed
- Reference (pytorch) and ttnn implementation
- Optimization (e.g. fusion, deallocation of tensors) when not needed, sharding, model tracing
- Demo and validation using a public dataset
- Testing for perf and pcc

### Checklist
- [ ] [All post commit](https://github.com/tenstorrent/tt-metal/actions/workflows/all-post-commit-workflows.yaml) CI passes
- [ ] [Blackhole Post commit](https://github.com/tenstorrent/tt-metal/actions/workflows/blackhole-post-commit.yaml) CI with demo tests passes (if applicable)
- [ ] [Model regression](https://github.com/tenstorrent/tt-metal/actions/workflows/perf-models.yaml) CI passes (if applicable)
- [ ] [Device performance regression](https://github.com/tenstorrent/tt-metal/actions/workflows/perf-device-models.yaml) CI passes (if applicable)
- [ ] (For models and ops writers) [Single-card demo tests](https://github.com/tenstorrent/tt-metal/actions/workflows/single-card-demo-tests.yaml) CI passes (if applicable)
- [ ] [Galaxy quick](https://github.com/tenstorrent/tt-metal/actions/workflows/galaxy-quick.yaml) CI passes (if applicable)
- [ ] [Galaxy demo tests, for Llama](https://github.com/tenstorrent/tt-metal/actions/workflows/galaxy-demo-tests.yaml) CI passes, if applicable, because of current Llama work
- [ ] (For runtime and ops writers) [T3000 unit tests](https://github.com/tenstorrent/tt-metal/actions/workflows/t3000-unit-tests.yaml) CI passes (if applicable, since this is run on push to main)
- [ ] (For models and ops writers) [T3000 demo tests](https://github.com/tenstorrent/tt-metal/actions/workflows/t3000-demo-tests.yaml) CI passes (if applicable, since this is required for release)
- [ ] [cpp-unit-tests](https://github.com/tenstorrent/tt-metal/actions/workflows/tt-metal-l2-nightly.yaml) job passes
- [ ] New/Existing tests provide coverage for changes

---
- To see the specific tasks where the Asana app for GitHub is being used, see below:
  - https://app.asana.com/0/0/1212832436210770

mert-kurttutan committed:
Unet3d model bring up start: working ttnn/reference impl and passing tests

mert-kurttutan committed:
Add more documentation of ttnn modules

mert-kurttutan committed:
Main parts for the unet3d bringup is complete
- Model with optimization
- Model tracing
- Validation on a dataset
- Testing and comparison against pytorch for each module

mert-kurttutan commented:
Below, I also attach the perf sheet and validation log that includes the validation result produced by validate script
[ops_perf_results_unet3d_2026_01_04_20_07_02.csv](https://github.com/user-attachments/files/24424246/ops_perf_results_unet3d_2026_01_04_20_07_02.csv)

[validation.log](https://github.com/user-attachments/files/24424253/validation.log)

mert-kurttutan committed:
3D DenseUNet bring up using TTNN APIs
- Added References to README.md

mert-kurttutan committed:
3D DenseUNet bring up using TTNN APIs
- Remove hardcoding of path for config file

mert-kurttutan committed:
3D DenseUNet bring up using TTNN APIs
- Add scikit to demo deps
- Fix the name of perf tests

mert-kurttutan committed:
3D DenseUNet bring up using TTNN APIs
- BEtter naming
- ensure no undefined variable for bias_tensor

mert-kurttutan committed:
3D DenseUNet bring up using TTNN APIs
- Get rid of old name for perf test func

mert-kurttutan committed:
3D DenseUNet bring up using TTNN APIs
- Fix some typo and uunused vars

mert-kurttutan commented:
@cglagovichTT Could you help with the review ?

mert-kurttutan committed:
3D DenseUNet bring up using TTNN APIs
- Remove hard coded device num for validation

mert-kurttutan committed:
3D DenseUNet bring up using TTNN APIs
- Remove some redundant commetns

mert-kurttutan committed:
3D DenseUNet bring up using TTNN APIs
- Remove final_sigmoid field since we always use sigmoid

yieldthought reviewed (approved):
I think this is looking good - do take a look at the codex comments. Approving to unblock!

mert-kurttutan committed:
3D DenseUNet bring up using TTNN APIs
- Fix halo shape logic error in the test_infra
- save prediction output during validation

mert-kurttutan commented:
@jberkowitzTT The error regarding csv file is an internal error [1]. It essentially does not put the "branch line" when it checks out from detached head (which my PR request is, as is every PR request). The corresponding logic is [here](https://github.com/tenstorrent/tt-metal/blob/f1f5abc8598426186251ccb7b59fd8ca0a55b1c7/models/perf/perf_utils.py#L30). This missing line leads to early stop error for the next function.


1: The csv file is supposed to be an output of the following internal command
```shell
pytest -n auto --timeout 300 models/experimental/stable_diffusion_xl_base/tests/test_sdxl_op_unit_test_perf.py -k "_performance" -xv --tb=short
```

dvartaniansTT reviewed (changes requested):
please include a polished version of such details and workaround in the README file. 

"This is the start of my PR. Regarding model implementations, since the model is 3D, it has some ops that is implemented natively, e.g. max_pool3d or upsample. For these, I implemented a temporary solution by using 2d version twice where you apply ops on along h,w dimensions and treat t dimension as batch. Then, you combine h w dimensions into a single dimension and treat t dimension as another dim. There seems to be 3 options:

Use 2d version
Use torch version as a fallback
And/or open issue for the native 3d implementation of these ops."

mert-kurttutan committed:
3D DenseUNet bring up using TTNN APIs
- Include workaround details in README.md

mert-kurttutan committed:
3D DenseUNet bring up using TTNN APIs
- Add info about the input dimension size

fvranicTT commented:
@mert-kurttutan could you please rebase against the latest main? AI Assistant error should go away.

dvartaniansTT committed:
Merge branch 'main' into unet3d
